### PR TITLE
test(e2e): fix retry on create konnect RG

### LINKED ISSUE #4193: Retry on creating Konnect RG causes flaky e2e tests
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

Retry on creating Konnect runtime groups causes flakiness on e2e test cases with konnect:
- https://github.com/Kong/kubernetes-ingress-controller/actions/runs/5307845863/jobs/9606756215
- https://github.com/Kong/kubernetes-ingress-controller/actions/runs/5301793857/jobs/9596305428

The reason is that the client times out on the first try of creating an RG, but the RG is acutually created on Konnect. When the client retries to create the same RG, Konnect returns 409 and the whole test failes. 

### Proposed Solution

- Update the logic of creating Konnect RG: first check if the RG exists, if the RG exists we either:
  - delete the existing one and create a new one
  - use the existing one

### Additional information

_No response_

### Acceptance Criteria

- [ ] e2e test cases does not fail because of duplicate creation of Konnect RG

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

current retry logic on creating test konnect RGs will fail when the RG is created, but the request timed out to get response. This caused e2e test cases fail when creating RGs on the situations above.

This PR checks for existence of RG with specified name before sending the creation request. 

**Which issue this PR fixes**:

fixes #4193 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: test/e2e/konnect_test.go [hunk 1]
```diff
@@ -190,28 +192,65 @@ func createTestRuntimeGroup(ctx context.Context, t *testing.T) string {
...
+		}
+		if listRgResp.StatusCode() != http.StatusOK {
+			return fmt.Errorf("failed to list RGs: code %d, message %s", listRgResp.StatusCode(), string(listRgResp.Body))
+		}
+
+		if listRgResp.JSON200 != nil && listRgResp.JSON200.Data != nil {
+			// REVIEW: what to do if the runtime group already exists?
+			// considering we are using a random UUID as runtime group name,
+			// the probability of conflict with other existing RGs is very small,
+			// so we use the RG here.
```
Comment (pmalek) [inline test/e2e/konnect_test.go:216]: I'd say not reuse it and fall back regenerate the name and try again with that. Someone using that already existing RG might be surprised that it got deleted after our tests complete.

BTW: is there a way to add labels or tags to those RGsS

Comment (randmonkey) [reply test/e2e/konnect_test.go]: Updated. Now it creates an RG with new name on failures, and delete all RGs created in the function in Cleanup.

File: test/e2e/konnect_test.go [hunk 2]
```diff
@@ -190,28 +192,79 @@ func createTestRuntimeGroup(ctx context.Context, t *testing.T) string {
...
+		// that RG created, but we timed out on getting the response.
+		// This prevents RG leaks.
+		listRgResp, err := rgClient.ListRuntimeGroupsWithResponse(ctx, &rg.ListRuntimeGroupsParams{})
+		assert.NoError(t, err, "failed to list runtime groups")
+		assert.Equalf(t, http.StatusOK, listRgResp.StatusCode(), "failed to list runtime groups: code %d, body %s",
+			listRgResp.StatusCode(), string(listRgResp.Body))
+		assert.NotNil(t, listRgResp.JSON200, "no body in list RG response")
+		if listRgResp.JSON200.Data == nil {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply test/e2e/konnect_test.go]: Added retry on deleting RGs.

File: test/e2e/konnect_test.go [hunk 3]
```diff
@@ -190,28 +192,113 @@ func createTestRuntimeGroup(ctx context.Context, t *testing.T) string {
...
 	require.NoError(t, err)
+	createTimeStampStr := fmt.Sprintf("%d", time.Now().Unix())
+	createTestUUID := uuid.NewString()
+
+	var rgID uuid.UUID
+
+	createRgErr := retry.Do(func() error {
+		rgName := uuid.NewString()
+		createRgResp, err := rgClient.CreateRuntimeGroupWithResponse(ctx, rg.CreateRuntimeGroupRequest{
+			Description: lo.ToPtr("This is a description"),
```
Comment (pmalek) [inline test/e2e/konnect_test.go:205]: We might want to use the description to add a name of the test running here via `t.Name`. OTOH this is the only test that uses it so it would be better to include something like GITHUB CI run ID if present in the env.

We could use `GITHUB_SERVER_URL`, `GITHUB_SERVER_URL` and `GITHUB_RUN_ID` from https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables

This way if we encounter any errors and we're looking at the problematic RG in konnect we can easily link those 2 (RG with E2E run).

File: test/e2e/konnect_test.go [hunk 4]
```diff
@@ -190,28 +192,113 @@ func createTestRuntimeGroup(ctx context.Context, t *testing.T) string {
...
+						_, deleteErr := rgClient.DeleteRuntimeGroupWithResponse(ctx, deleteRgID)
+						if deleteErr != nil {
+							t.Logf("failed to delete runtime group %s: %v", deleteRgID, err)
+							deleteRgErrors = append(deleteRgErrors, deleteErr)
+						}
+					}
+				}
+				return errors.Join(deleteRgErrors...)
+			},
+			retry.Attempts(5), retry.Delay(time.Second),
```
Comment (pmalek) [inline test/e2e/konnect_test.go:295]: This whole section responsible for deleting RGs could be a function to increase the  readability of this section. WDYT?

File: test/e2e/konnect_test.go (hunk 3 above)
Comment (randmonkey) [reply test/e2e/konnect_test.go]: done.

File: test/e2e/konnect_test.go (hunk 4 above)
Comment (randmonkey) [reply test/e2e/konnect_test.go]: extracted the procedure to another function.

File: test/e2e/konnect_test.go [hunk 5]
```diff
@@ -190,28 +263,73 @@ func createTestRuntimeGroup(ctx context.Context, t *testing.T) string {
...
+		// This prevents RG leaks.
+		t.Logf("deleting other runtime groups created in the test")
+		err = retry.Do(
+			deleteRuntimeGroupsCreatedInTest(
+				ctx, *rgClient,
+				t, createTimeStamp, createTestUUID,
+			),
+			retry.Attempts(5), retry.Delay(time.Second),
+		)
+		assert.NoError(t, err, "failed to list and delete other runtime groups created in the test")
```
Comment (czeslavo) [inline test/e2e/konnect_test.go:328]: Given we will have the cleanup job taking care of the orphaned runtime groups every 30 minutes, I would suggest dropping the `deleteRuntimeGroupsCreatedInTest` and `create_test_timestamp`, `create_test_uuid` labels added to the runtime group. 

IMO it's not worth taking care of every possible edge case here as anyway we can't be sure that something won't go wrong with deleting. We could assume that the test does not create more than one RG and handle only the "happy path" in the test (APIs are responding with no timeouts which should be the case if we will cleanup runtime groups' roles as in https://github.com/Kong/kubernetes-ingress-controller/pull/4226), leaving the edge cases to be handled by the asynchronous job. If anything goes wrong and we create more than 1 RG, it will get pruned in an hour.

**Review by czeslavo — COMMENTED** [review]
I suggest to limit this PR's changes to improving the runtime group's description which will improve visibility in case something goes really wrong. Given the asynchronous cleanup job (implemented in https://github.com/Kong/kubernetes-ingress-controller/pull/4226) will handle orphaned RGs, I think it will be better to keep the test suite code as simple as possible, WDYT?

File: test/e2e/konnect_test.go (hunk 5 above)
Comment (randmonkey) [reply test/e2e/konnect_test.go:328]: Yes, I think it is unnecessary to delete the "leaked" RGs when we have the cleanup job to remove them.

**Review by czeslavo — APPROVED** [review]

**Comment by randmonkey** [comment]
@czeslavo Please look at the PR again after I resolved the conflicts to see if there are any unexpected things on the merge.

**Comment by czeslavo** [comment]
I added `ci/run-e2e` again to ensure it works just in case. LGTM though 👍

**Comment by randmonkey** [comment]
@czeslavo https://github.com/Kong/kubernetes-ingress-controller/actions/runs/5387501095 finished and succeeded.

**Review by czeslavo — APPROVED** [review]
